[WIP] Fonts (#1039)
withastro/roadmap: proposals/0052-fonts.md

# Summary

Have first-party support for fonts in Astro.

# Example

```js
// astro config
export default defineConfig({
  fonts: {
    families: ["Roboto", "Lato"],
  },
});
```

```astro
---
// layouts/Layout.astro
import { Font } from 'astro:fonts'
---
<head>
	<Font family='Inter' preload />
	<Font family='Lato' />
	<style>
		h1 {
			font-family: var(--astro-font-inter);
		}
		p {
			font-family: var(--astro-font-lato);
		}
	</style>
</head>
```

# Background & Motivation

Fonts is one of those basic things when making a website, but also an annoying one to deal with. Should I just use a link to a remote font? Or download it locally? How should I handle preloads then?

The goal is to improve the DX around using fonts in Astro.

> Why not using fontsource?

Fontsource is great! But it's not intuitive to preload, and more importantly, doesn't have all fonts. The goal is to have a more generic API for fonts (eg. you want to use a paid provider like adobe).

# Goals

- Specify what font to use
- Cache fonts
- Specify what provider to use
- Load/preload font on a font basis
- Generate fallbacks automatically
- Performant defaults

> - Runtime agnostic

Author: ```suggestion
- Runtime agnostic
- Configure font families (subset, unicode range, weights etc)
```

# Non-Goals

- Runtime API (SSR is supported tho)

> - Automatic subsetting

Author: ```suggestion
- Automatic subsetting (eg. analyzing static content)
```

**If you have feedback and the feature is released as experimental, please leave it on the Stage 3 PR. Otherwise, comment on the Stage 2 issue (links below).**

- Start Date: 2024-10-15
- Reference Issues: <!-- related issues, otherwise leave empty -->
- Implementation PR: <!-- leave empty -->
- Stage 2 Issue: https://github.com/withastro/roadmap/issues/1037
- Stage 3 PR: https://github.com/withastro/roadmap/pull/1039

# Summary

Have first-party support for fonts in Astro.

# Example

```js
// astro config
export default defineConfig({
  fonts: {
    families: ["Roboto", "Lato"],
  },
});
```

```astro
---
// layouts/Layout.astro
import { Font } from 'astro:fonts'
---
<head>
	<Font family='Inter' preload />
	<Font family='Lato' />
	<style>
		h1 {

> 			font-family: var(--astro-font-inter);

Author: Well since it's just some css, I don't think it will cause any issues? In Tailwind 3, you'd have to reference like so:
```js
const defaultTheme = require('tailwindcss/defaultTheme')

module.exports = {
  theme: {
    extend: {
      fontFamily: {
        'sans': ['var(--astro-font-inter)', ...defaultTheme.fontFamily.sans],
      },
    }
  }
}
```

> 			font-family: var(--astro-font-inter);

Reviewer: Maybe allowthe variable name to be configured in the tag, i.e. `<Font family='Lato' variable='my-custom-font' />` which generates `var(--my-custom-font)`?

This is similar to how [Next.js does it](https://nextjs.org/docs/pages/building-your-application/optimizing/fonts#with-tailwind-css), albeit they use a function syntax instead of a JSX-style component

		}
		p {
			font-family: var(--astro-font-lato);
		}
	</style>
</head>
```

# Background & Motivation

Fonts is one of those basic things when making a website, but also an annoying one to deal with. Should I just use a link to a remote font? Or download it locally? How should I handle preloads then?

The goal is to improve the DX around using fonts in Astro.

> Why not using fontsource?

Fontsource is great! But it's not intuitive to preload, and more importantly, doesn't have all fonts. The goal is to have a more generic API for fonts (eg. you want to use a paid provider like adobe).

# Goals

- Specify what font to use
- Cache fonts
- Specify what provider to use
- Load/preload font on a font basis
- Generate fallbacks automatically
- Performant defaults
- Runtime agnostic
- Configure font families (subset, unicode range, weights etc)

# Non-Goals

- Runtime API (SSR is supported tho)
- Automatic subsetting (eg. analyzing static content)

> - Font detection

Reviewer 2: Would be great to specify what this means, I know it personally but people might not!

# Detailed Design

## Astro config

### Overview

The goal is to have a config that starts really simple for basic usecases, but can also be complex for advanced usecases. Here's an example of basic config:

```js
import { defineConfig } from "astro/config";

export default defineConfig({
  fonts: {
    families: ["Roboto", "Lato"],
  },
});
```

That would get fonts from [Google Fonts](https://fonts.google.com/) with sensible defaults.

Here's a more complex example:

```js
import { defineConfig, fontProviders } from "astro/config";
import { myCustomFontProvider } from "./provider";

export default defineConfig({
  fonts: {
    providers: [
      fontProviders.adobe({ apiKey: process.env.ADOBE_FONTS_API_KEY }),
      myCustomFontProvider(),
    ],

>     defaults: {

Reviewer 2: I think `defaults` make this a nicer API overall, but I wonder how required it is, do people really have that much fonts that setting defaults is required 🤔

Author: I have to admit I rarely use a lot of fonts but I guess that could be the case for more complex sites. IMO people will probably want this for setting a default provider in most cases, so why not support all the other properties I suppose!

>       provider: "adobe",

Reviewer 2: how would you know the name of the provider here? Will it be... typed (👀) automatically based on what you put in `providers`?

Author: Yes! Thanks to `defineConfig()` being generic now, we can do a lot of cool stuff here. Providers would be easily typed

      weights: [200, 700],
      styles: ["italic"],

>       subsets: [

Reviewer 2: would love an example with opentype fonts features

Author: This is not a subet right? I guess that would need to be more keys, something like the following:
```js
{
	fontKerning: "'kern' 1",
	fontVariantAlternates: "1"
	// etc...
}
```

MDN docs: https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_fonts/OpenType_fonts_guide

        "cyrillic-ext",
        "cyrillic",
        "greek-ext",
        "greek",
        "vietnamese",
        "latin-ext",
        "latin",
      ],
    },
    families: [
      "Roboto",
      {
        name: "Lato",
        provider: "google",
        weights: [100, 200, 300],
      },
      {
        name: "Custom",
        provider: "local",
        src: ["./assets/fonts/Custom.woff2"],
      },
    ],
  },
});
```

### Providers

#### Definition

> A provider allows to retrieve font faces data from a font family name from a given CDN or abstraction. It's a [unifont](https://github.com/unjs/unifont) provider.

Reviewer 2: I think some people could have a concern with us exposing a third-party "API" like that in the config, but we already do for syntax highlighting themes, so I think it's okay. Plus, the default providers unifont provide are probably enough for 99.999% of people anyway.

##### Google

This is the default, and it's not configurable. Given the amount of fonts it supports by, it sounds like a logic choice. Note that the default can be customized for more advanced usecases.

```js
export default defineConfig({
  fonts: {
    families: ["Roboto"],
  },
});
```

```js
export default defineConfig({
  fonts: {
    defaults: {
      provider: "local",
    },
    families: [
      {
        name: "Roboto",
        provider: "google",
      },
    ],
  },
});
```

##### Local

This provider, unlike all the others, requires paths to fonts relatively to the root.

```js
import { defineConfig, fontProviders } from "astro/config";

export default defineConfig({
  fonts: {
    families: [
      {
        name: "Custom",
        provider: "local",
        src: ["./assets/fonts/Custom.woff2"],
      },
    ],
  },
});
```

#### Opt-in providers

Other unifont providers are exported from `astro/config`.

```js
import { defineConfig, fontProviders } from "astro/config";

export default defineConfig({
  fonts: {
    providers: [
      fontProviders.adobe({ apiKey: process.env.ADOBE_FONTS_API_KEY }),
    ],
    // ...
  },
});
```

#### Why this API?

> 1. **Coherent API**: a few things in Astro are using this pattern, namely integrations and vite plugins. It's simple to author as a library author, easy to use as a user

Reviewer 3: I disagree with this statement, this API is used **only** for integrations and vite plugins. 

For example: middleware, actions, and content collections (we could consider them an exception though) don't follow this pattern. They use a **known file** to collect users' APIs. Same thing for `@astrojs/db` (but that's integration).

I think that our configuration should inject as less runtime as possible, use good defaults, and maybe the custom providers could follow the same pattern we use with actions and middleware.

Reviewer 2: Image services, which are probably the most similar thing do work this way

Reviewer 3: Probably I missed something, please correct me if I misunderstood what you mean. Based on this docs page (https://docs.astro.build/en/reference/image-service-reference/#user-configuration), the image service uses an entry point API, which is different from the one proposed.

Reviewer 2: Ah, apologies, I misunderstood what you meant. Thought this was mostly about where they're configured.

Image services do use an entrypoint, but they're configured in the Astro config, when they have settings.

Author: The main advantage of having objects in the config directly for providers is typing support, which imo is super important

Author: Alright so I guess you'd prefer such an API, am I right?
```js
defineConfig({
	fonts: {
		providers: [fontProviders.adobe({ apiKey: "..." })]
		// equal to { name: "adobe", entrypoint: "astro/assets/fonts/adobe", config: { apiKey: "..." } }
	}
})
```
I guess that's doable

2. **Keep opt-in providers**: allows to only use 2 providers by default, and keeps the API open to anyone
3. **Types!**: now that `defineConfig` [supports generics](https://github.com/withastro/astro/pull/12243), we can do powerful things! Associated with type generation, we can generate types for `families` `name`, infer the provider type from `defaults.provider` and more.

### Defaults

Astro must provide sensible defaults when it comes to font weights, subsets and more. But when dealing with more custom advanced setups, it makes sense to be able to customize those defaults. They can be set in `fonts.defaults` and will be merged with Astro defaults.

We need to decide what default to provide. I can see 2 paths:

| Path      | Example (weight)    | Advantage             | Downside                                                                          |
| --------- | ------------------- | --------------------- | --------------------------------------------------------------------------------- |

> | Minimal   | Only include `400`  | Lightweight           | People will probably struggle by expecting all weights to be available by default |

Reviewer 2: I'm team just 400, Google Fonts also gives you only 400 when you unselect the "give all options" setting

```astro
---
import { Font } from "astro:assets"
---
<head>
  <Font family="Inter" preload cssVar="primary-font" />
  <Font family="Lato" />
</head>
```

### Family

The family will be typed using type gen, based on the user's config.

### Preload

Defaults to `false`:

- **Enabled**: Outputs a preload link tag and a style tag, without fallbacks
- **Disabled**: Output a style tag with fallbacks (generated using [fontaine](https://github.com/unjs/fontaine))

### cssVar

Defaults to `astro-font-${computedFontName}`. Specifies what identifier to use for the generated css variable. This is useful for font families names that may contain special character or conflict with other fonts.

## Usage

Since fallbacks may be generated for a given family name, this name can't be used alone reliably:

```css
h1 {
  font-family: "Inter"; /* Should actually be "Inter", "Inter Fallback" */
}
```

To solve this issue, a css variable is provided by the style tage generated by the `<Font />` component:

```css
h1 {
  font-family: var(--astro-font-inter); /* "Inter", "Inter Fallback" */
}
```

## Fallbacks

> We try to generate useful fallbacks when possible (and if the user specifies at least one):

Author: Sure I'll list the conditions required

- Using `fontaine`, we get metrics using the family name
- We check the last fallback to see if it's a generic family name (eg. sans-serif)
- We get associated local fonts (eg. Arial)
- We generate font faces so that the local font matches the original font shape

The easiest way to benefit from fallback generation is by doing the following:

```js
{
  family: 'Roboto',
  fallbacks: ['sans-serif']
}
```

> This will give `Roboto, 'Roboto fallback: Arial', sans-serif`. This way we can infer what local font to use. It's still possible to provide more fallbacks:

Author: This is what we generate

```js
{
  family: 'Roboto',
  fallbacks: ['Times New Roman', 'sans-serif']
}
```

This will give `Roboto, 'Roboto fallback: Arial', 'Times New Roman', sans-serif`.

## How it works under the hood

- Once the config is fully resolved, we get fonts face data using `unifont`
- We generate fallbacks using `fontaine` and pass all the data we need through a virtual import, used by the `<Font />` component
- We inject a vite middleware in development to download fonts as they are requested in development

> - During build, we download all fonts and put them in `outDir`

Author: This is outdated

- Stage 2 Issue: https://github.com/withastro/roadmap/issues/1037
- Stage 3 PR: https://github.com/withastro/roadmap/pull/1039

# Summary

Have first-party support for fonts in Astro.

# Example

```js
// astro config
export default defineConfig({
  fonts: {
    families: ["Roboto", "Lato"],
  },
});
```

```astro
---
// layouts/Layout.astro
import { Font } from 'astro:fonts'
---
<head>
	<Font family='Inter' preload />
	<Font family='Lato' />
	<style>
		h1 {
			font-family: var(--astro-font-inter);
		}
		p {
			font-family: var(--astro-font-lato);
		}
	</style>
</head>
```

# Background & Motivation

Fonts is one of those basic things when making a website, but also an annoying one to deal with. Should I just use a link to a remote font? Or download it locally? How should I handle preloads then?

The goal is to improve the DX around using fonts in Astro.

> Why not using fontsource?

Fontsource is great! But it's not intuitive to preload, and more importantly, doesn't have all fonts. The goal is to have a more generic API for fonts (eg. you want to use a paid provider like adobe).

# Goals

- Specify what font to use
- Cache fonts

> - Specify what provider to use

Reviewer 4: Can you define provider somewhere?

Reviewer 4: Oh sorry, I see that you do that below, nevermind.

- Load/preload font on a font basis

> - Generate fallbacks automatically

Reviewer 4: is this configurable? should it be?

Author: It is not configurable atm. If we can generate one we will and it will take precedence over the fallbacks defined by the user. I guess we could make it configurable, what name do you think would be good? `automaticFallback: false`?

| Minimal   | Only include `400`  | Lightweight           | People will probably struggle by expecting all weights to be available by default |
| Extensive | Include all weights | Predictable for users | Heavier by default                                                                |

### Families

A family is made of a least a `name`:

```js
export default defineConfig({
  fonts: {
    families: [
      {
        name: "Roboto",
      },
      "Roboto", // Shorthand
    ],
  },
});
```

It can specify options such as `weights` and `subsets`:

```js
export default defineConfig({
  fonts: {
    families: [
      {
        name: "Roboto",
        weights: [400, 600],
      },
    ],
  },
});
```

It can also specify a `provider` (and `src` if it's the `local` provider):

```js
export default defineConfig({
  fonts: {
    families: [
      {
        name: "Roboto",
        provider: "local",
        src: "./Roboto.woff2",
      },
    ],
  },
});
```

It also accepts a `as` prop, which will be used when generating CSS. This is useful in the following cases:

> 1. You are an integration and want to make sure you don't conflict with fonts declared by the user
> 2. Your font name can't be properly converted to a CSS variable name

Reviewer 3: Who is "you" in these sentences? I think we should rephrase them:

```suggestion
1. Avoid conflicts between the font names declared by integrations and those declared by the end-users.
2. A font name can't be properly converted to a CSS variable name.
```

As for the second bullet point, we should at least provide an example of when a font can't be converted to a CSS variable.

Author: I agree with improving the wording! As for point 2, I haven't experienced it but I was thinking about non latin characters (eg. chinese): are they allowed as a css variable name?

Author: I found https://www.thingsaboutweb.dev/en/posts/css-variables-part-1 but I'm still unsure. I think there are cases where the variable name contains characters that will be stripped out and `as` will allow the user to have more control

Author: We are already using `cssesc` for view transitions, i think we could reuse it to take care of encoding
